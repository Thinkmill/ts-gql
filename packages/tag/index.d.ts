import { DocumentNode } from "graphql";

// TODO: subscriptions

export type BaseTypedDocument = {
  result: any;
  documents: Record<string, TypedDocumentNode<BaseDocumentTypes>>;
  fragments: (a: any) => any;
};

export type BaseTypedQuery = BaseTypedDocument & {
  type: "query";
  variables: any;
};

export type BaseTypedMutation = BaseTypedDocument & {
  type: "mutation";
  variables: any;
};

export type BaseOperations = BaseTypedQuery | BaseTypedMutation;

export type BaseTypedFragment = BaseTypedDocument & {
  name: string;
  type: "fragment";
};

export type BaseDocumentTypes = BaseOperations | BaseTypedFragment;

export type TypedDocumentNode<
  TypedDocument extends BaseDocumentTypes
> = DocumentNode & {
  ___type: TypedDocument;
} & (TypedDocument extends BaseOperations
    ? import("@graphql-typed-document-node/core").TypedDocumentNode<
        TypedDocument["result"],
        TypedDocument["variables"]
      >
    : unknown);

export type OperationData<
  Node extends TypedDocumentNode<BaseOperations>
> = Node["___type"]["result"];

export type OperationVariables<
  Node extends TypedDocumentNode<BaseOperations>
> = Node["___type"]["variables"];

export type FragmentData<
  Node extends TypedDocumentNode<BaseTypedFragment>
> = Node["___type"]["result"];

export type AllDocuments<
  Node extends TypedDocumentNode<BaseDocumentTypes>
> = Node["___type"]["documents"];

export declare function gql(strings: TemplateStringsArray): never;

/**
 * @deprecated {@linkcode TypedDocumentNode} is now assignable to {@linkcode DocumentNode}
 * so this call can be removed and the {@linkcode TypedDocumentNode} can be used directly
 */
export function getDocumentNode(
  node: TypedDocumentNode<BaseDocumentTypes>
): DocumentNode;
