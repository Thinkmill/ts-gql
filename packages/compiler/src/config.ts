import path from "path";
import { findPkgJsonFieldUp } from "find-pkg-json-field-up";
import { getSchemaFromOptions } from "./get-schema";

export async function getRawConfig(cwd: string) {
  let { packageJson, directory } = await findPkgJsonFieldUp("ts-gql", cwd);
  let field = packageJson["ts-gql"];
  if (
    typeof field === "object" &&
    field !== null &&
    typeof field.schema === "string"
  ) {
    return {
      schema: path.resolve(directory, field.schema),
      directory,
    };
  }
  throw new Error("Config not found");
}

export async function getConfig(cwd: string) {
  let config = await getRawConfig(cwd);
  return {
    directory: config.directory,
    schema: await getSchemaFromOptions(config.schema),
  };
}
