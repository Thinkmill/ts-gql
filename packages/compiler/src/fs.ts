import gracefulFS from "graceful-fs";
import { promisify } from "util";

export const writeFile = promisify(gracefulFS.writeFile);
export const readFile = promisify(gracefulFS.readFile);
export const mkdir = promisify(gracefulFS.mkdir);
export const readdir = promisify(gracefulFS.readdir);
export const unlink = promisify(gracefulFS.unlink);
