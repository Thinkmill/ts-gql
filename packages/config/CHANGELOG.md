# @ts-gql/config

## 0.5.0

### Minor Changes

- [`fba7341`](https://github.com/Thinkmill/ts-gql/commit/fba7341a1418e0a9d555172dc5c6e86899fa6ed3) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add `readonlyTypes` option

## 0.4.0

### Minor Changes

- [`4be8faa`](https://github.com/Thinkmill/ts-gql/commit/4be8faafa0fba17efa491a0aec8ddbb472aa5572) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Replace `nonOptionalTypename` option with `addTypename` option

## 0.3.0

### Minor Changes

- [`d798897`](https://github.com/Thinkmill/ts-gql/commit/d7988972e801c41bb96aaa4dec5763ebae73e30e) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add nonOptionalTypename option

## 0.2.0

### Minor Changes

- [`caa1974`](https://github.com/Thinkmill/ts-gql/commit/caa19743de1aa1345795691b8d4eea58c052fc8f) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support scalars option

## 0.1.0

### Minor Changes

- [`e42383b`](https://github.com/Thinkmill/ts-gql/commit/e42383b5970a554462384f9851aabc173f7fcf52) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Initial Release
