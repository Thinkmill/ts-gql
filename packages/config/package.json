{
  "name": "@ts-gql/config",
  "version": "0.7.1",
  "main": "dist/config.cjs.js",
  "module": "dist/config.esm.js",
  "license": "MIT",
  "dependencies": {
    "@babel/runtime": "^7.9.2",
    "find-pkg-json-field-up": "^1.0.1",
    "lazy-require.macro": "^0.1.0",
    "superstruct": "^0.10.12"
  },
  "peerDependencies": {
    "graphql": "^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0 || ^15.0.0"
  },
  "devDependencies": {
    "graphql": "^14.0.0"
  },
  "repository": "https://github.com/Thinkmill/ts-gql/tree/master/packages/config"
}
