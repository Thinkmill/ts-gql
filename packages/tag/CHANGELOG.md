# @ts-gql/tag

## 0.7.0

### Minor Changes

- [#98](https://github.com/Thinkmill/ts-gql/pull/98) [`bc8d6eb`](https://github.com/Thinkmill/ts-gql/commit/bc8d6ebbf1021829de24d3c916dad5e0b3ab1edf) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - `ts-gql`'s `TypedDocumentNode` type is now compatible with [`@graphql-typed-document-node/core`](https://github.com/dotansimha/graphql-typed-document-node)'s `TypedDocumentNode`.

  The recommended usage of ts-gql with Apollo Client is now to use `@apollo/client` directly. This also allows ts-gql to be used with urql and any other GraphQL client that supports `@graphql-typed-document-node/core`. The `@ts-gql/apollo` package can still be used and may be updated in the future to avoid breakage if that makes sense but it is no longer the recommended pattern.

  When using `@apollo/client` over `@ts-gql/apollo`, it's important to note that some type safety will be lost:

  - Variables are always optional so omitting variables when they are required will no longer be caught by TypeScript
  - `refetchQueries` will accept any string so passing names to queries that don't exist will not cause a TypeScript error. You should likely pass in the document with the query itself to avoid mis-typing query names causing errors.

  Because `@graphql-typed-document-node/core`'s `TypedDocumentNode` extends `graphql`'s `DocumentNode`, this means that `getDocumentNode` from `@ts-gql/tag` is no longer necessary. This could be another cause for bugs if there are two APIs, one that accepts a `TypedDocumentNode` that you should use and another that accepts `DocumentNode` which you shouldn't use, you could accidentally use the API that accepts `DocumentNode` over the one that accepts `TypedDocumentNode` where previously you would get an error when passing a `TypedDocumentNode` to something accepting a `DocumentNode`.

  #### Context behind this change

  When ts-gql was originally written, `@graphql-typed-document-node/core` did not exist. Since then, `@graphql-typed-document-node/core` has become used by Apollo Client and urql. Given that, maintaining types to adapt Apollo Client to ts-gql's `TypedDocumentNode` seems less sensible.

  While this does mean that some of ts-gql's safety is reduced, this seems like an appropriate trade-off so that ts-gql can reduce maintaince burden, avoid imposing opinions on top of GraphQL clients and support more GraphQL clients without having to write types for them specifically.

## 0.6.1

### Patch Changes

- [`0e3e2f5`](https://github.com/Thinkmill/ts-gql/commit/0e3e2f5004c7e42bbc394664c5e667ce3597e6fd) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support `graphql@16`

## 0.6.0

### Minor Changes

- [#90](https://github.com/Thinkmill/ts-gql/pull/90) [`dc22e45`](https://github.com/Thinkmill/ts-gql/commit/dc22e457d14c816274037010a627d10bcb30f11d) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Removed unused `Documents` export.

* [#90](https://github.com/Thinkmill/ts-gql/pull/90) [`dc22e45`](https://github.com/Thinkmill/ts-gql/commit/dc22e457d14c816274037010a627d10bcb30f11d) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Added `name` property to `BaseTypedFragment` and `fragments` property to `BaseTypedDocument`

- [#90](https://github.com/Thinkmill/ts-gql/pull/90) [`dc22e45`](https://github.com/Thinkmill/ts-gql/commit/dc22e457d14c816274037010a627d10bcb30f11d) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Added support for `"mode": "no-transform"` and `"mode": "mixed"`. See https://github.com/Thinkmill/ts-gql/blob/main/docs/no-transform.md for more details

* [#90](https://github.com/Thinkmill/ts-gql/pull/90) [`dc22e45`](https://github.com/Thinkmill/ts-gql/commit/dc22e457d14c816274037010a627d10bcb30f11d) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Removed spread parameter from `gql` export to align with the fact that `gql` from the `@ts-gql/tag` entrypoint does not allow interpolations.

## 0.5.3

### Patch Changes

- [`3149ffe`](https://github.com/Thinkmill/ts-gql/commit/3149ffe2ffb428273e80451d8a67873073e052c8) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support `graphql@^15.0.0`

## 0.5.2

### Patch Changes

- [`ab25d45`](https://github.com/Thinkmill/ts-gql/commit/ab25d45bd80dfe58f878a500c92e0bdb3eef5c86) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Improve error message when `gql` is called at runtime

## 0.5.1

### Patch Changes

- [`89bbe8b`](https://github.com/Thinkmill/ts-gql/commit/89bbe8bbd8c3ed3fd3d42ccb5fb0bfacb0d15575) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Fix the type of the `documents` field

## 0.5.0

### Minor Changes

- [`abba421`](https://github.com/Thinkmill/ts-gql/commit/abba4214b10bc878de9c7c9e350e5ef04f3ef11f) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Include all documents on typed document nodes

## 0.4.0

### Minor Changes

- [`b83e180`](https://github.com/Thinkmill/ts-gql/commit/b83e180ea94cd7fb1d66d5c7835f333a5fcf56f5) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Rename some types

## 0.3.0

### Minor Changes

- [`e4c60ad`](https://github.com/Thinkmill/ts-gql/commit/e4c60adcc45abba018c4b9d4d0379e7d529a9af1) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Use new technique to generate types.

  This requires you to use `@ts-gql/compiler` and `@ts-gql/babel-plugin` in addition to `@ts-gql/eslint-plugin`.

  Configuration also now lives in the `package.json` like this:

  ```json
  {
    "ts-gql": {
      "schema": "schema.graphql"
    }
  }
  ```

## 0.2.0

### Minor Changes

- [`8485b1a`](https://github.com/Thinkmill/ts-gql/commit/8485b1a28228feea836d076cc7dd1a0691414248) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Remove -with-required-variables types

## 0.1.0

### Minor Changes

- [`b444283`](https://github.com/Thinkmill/ts-gql/commit/b44428353e6e94f7df60b8ffc409b44b6fbca1ca) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Initial release
