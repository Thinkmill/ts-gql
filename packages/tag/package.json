{
  "name": "@ts-gql/tag",
  "version": "0.6.1",
  "main": "dist/tag.cjs.js",
  "module": "dist/tag.esm.js",
  "license": "MIT",
  "types": "index.d.ts",
  "dependencies": {
    "graphql-tag": "^2.12.6"
  },
  "peerDependencies": {
    "graphql": "^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0 || 14 || 15 || 16"
  },
  "devDependencies": {
    "graphql": "^16.3.0"
  },
  "repository": "https://github.com/Thinkmill/ts-gql/tree/main/packages/tag",
  "preconstruct": {
    "entrypoints": [
      "index.js",
      "no-transform.js"
    ]
  }
}