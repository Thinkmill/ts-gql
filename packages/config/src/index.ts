import path from "path";
import { GraphQLSchema } from "graphql";
import {
  findPkgJsonFieldUp,
  findPkgJsonFieldUpSync,
} from "find-pkg-json-field-up";
import fs from "fs";
import { promisify } from "util";
import { parseSchema, hashSchema } from "./read-schema";

export { parseSchema, hashSchema };

export class ConfigNotFoundError extends Error {}

export type Config = {
  directory: string;
  schema: GraphQLSchema;
  scalars: Record<string, string>;
  addTypename: boolean;
  readonlyTypes: boolean;
  schemaHash: string;
};

export type RawConfig = {
  directory: string;
  schema: string;
  scalars: Record<string, string>;
  addTypename: boolean;
  readonlyTypes: boolean;
};

function parseFieldToConfig({
  packageJson,
  directory,
}: {
  packageJson: Record<string, any>;
  directory: string;
}): RawConfig {
  let field = packageJson["ts-gql"];
  if (
    typeof field === "object" &&
    field !== null &&
    typeof field.schema === "string"
  ) {
    return {
      schema: path.resolve(directory, field.schema),
      directory,
      scalars: field.scalars || {},
      addTypename: field.addTypename ?? true,
      readonlyTypes: field.readonlyTypes ?? true,
    };
  }
  throw new ConfigNotFoundError("ts-gql config not found");
}

export async function getRawConfig(cwd: string) {
  return parseFieldToConfig(await findPkgJsonFieldUp("ts-gql", cwd));
}

export function getRawConfigSync(cwd: string) {
  return parseFieldToConfig(findPkgJsonFieldUpSync("ts-gql", cwd));
}

const readFile = promisify(fs.readFile);

export async function getConfig(cwd: string): Promise<Config> {
  let config = await getRawConfig(cwd);
  let schemaContents = await readFile(config.schema, "utf8");
  return {
    ...config,
    schemaHash: hashSchema(schemaContents),
    schema: parseSchema(config.schema, schemaContents),
  };
}

export function getConfigSync(cwd: string): Config {
  let config = getRawConfigSync(cwd);
  let schemaContents = fs.readFileSync(config.schema, "utf8");
  return {
    ...config,
    schemaHash: hashSchema(schemaContents),
    schema: parseSchema(config.schema, schemaContents),
  };
}
