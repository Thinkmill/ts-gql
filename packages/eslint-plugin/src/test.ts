import { SnapshotCreator } from "eslint-snapshot-test";
import { rules } from ".";
import { Config } from "@ts-gql/config";
import fs from "fs";
import { buildSchema } from "graphql";
import path from "path";
import { directory } from "tempy";
import { schema } from "../../compiler/src/test/test-schema";

let snapshotCreator = new SnapshotCreator({
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2020,
    ecmaFeatures: {
      jsx: true,
    },
    sourceType: "module",
  },
});

let builtSchema = buildSchema(schema);

let testConfig: Config = {
  addTypename: true,
  directory: __dirname,
  readonlyTypes: true,
  scalars: {},
  schemaHash: "123",
  schemaFilename: "schema.graphql",
  schema: () => builtSchema,
  mode: "transform",
};

let fixturesPath = path.join(__dirname, "__fixtures__");

let fixtures = fs.readdirSync(fixturesPath);

for (const fixture of fixtures) {
  test(fixture, async () => {
    const fixturePath = path.join(fixturesPath, fixture);
    const code = fs.readFileSync(fixturePath, "utf8");
    let result = snapshotCreator
      .mark({
        code,
        rule: rules["ts-gql"],
        ruleName: "ts-gql",
      })
      .withFileName(fixturePath)
      .withOptions([
        {
          ...testConfig,
          mode: fixture.endsWith(".no-transform.ts")
            ? "no-transform"
            : fixture.endsWith(".mixed.ts")
            ? "mixed"
            : "transform",
        },
      ])
      .render();
    if (result.fixedOutput === code) {
      delete result.fixedOutput;
    }
    expect(result).toMatchSnapshot();
  });
}

test("config and file in the same directory", () =>
  directory.task(async (tmpPath) => {
    const filePath = path.join(tmpPath, "test.ts");
    fs.writeFileSync(
      `${tmpPath}/package.json`,
      JSON.stringify({
        name: "blah",
        "ts-gql": {
          schema: "schema.graphql",
        },
      })
    );
    fs.writeFileSync(`${tmpPath}/schema.graphql`, schema);

    const code = `
    import { gql } from "@ts-gql/tag";
    gql\`
      query Something {
        hello
      }
    \`;
    `;
    fs.writeFileSync(filePath, code);
    let result = snapshotCreator
      .mark({
        code,
        rule: rules["ts-gql"],
        ruleName: "ts-gql",
      })
      .withFileName(filePath)
      .render();
    expect(result).toMatchInlineSnapshot(`
      {
        "fixedOutput": "
          import { gql } from "@ts-gql/tag";
          gql\`
            query Something {
              hello
            }
          \`as import("./__generated__/ts-gql/Something").type;
          ",
        "lintMessages": [
          {
            "column": 5,
            "endColumn": 6,
            "endLine": 7,
            "fix": {
              "range": [
                100,
                100,
              ],
              "text": "as import("./__generated__/ts-gql/Something").type",
            },
            "line": 3,
            "message": "You must cast gql tags with the generated type",
            "messageId": "mustUseAs",
            "nodeType": "TaggedTemplateExpression",
            "ruleId": "ts-gql",
            "severity": 2,
          },
        ],
        "snapshot": "

          import { gql } from "@ts-gql/tag";
          gql\`
          ~~~~    [You must cast gql tags with the generated type]
            query Something {
      ~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
              hello
      ~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
            }
      ~~~~~~~    [You must cast gql tags with the generated type]
          \`;
      ~~~~~    [You must cast gql tags with the generated type]
          ",
      }
    `);
  }));
