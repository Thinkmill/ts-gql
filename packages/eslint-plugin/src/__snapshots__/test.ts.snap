// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing {
    something
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "fix": Object {
        "range": Array [
          76,
          76,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing {
~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    something
~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`missing-id.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing {
    someObj {
      id
      other
    }
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 9,
      "fix": Object {
        "range": Array [
          94,
          94,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
    Object {
      "column": 14,
      "endColumn": 6,
      "endLine": 7,
      "fix": Object {
        "range": Array [
          70,
          70,
        ],
        "text": "
      id",
      },
      "line": 5,
      "message": "The id field on object and interface types must always be fetched if it exists so that caching works reliably",
      "messageId": "mustFetchId",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing {
~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    someObj {
~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
             ~    [The id field on object and interface types must always be fetched if it exists so that caching works reliably]
      other
~~~~~~~~~~~    [You must cast gql tags with the generated type]
~~~~~~~~~~~    [The id field on object and interface types must always be fetched if it exists so that caching works reliably]
    }
~~~~~    [You must cast gql tags with the generated type]
~~~~~    [The id field on object and interface types must always be fetched if it exists so that caching works reliably]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`missing-variable.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing($thing: String) {
    optional(thing: $thing)
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "fix": Object {
        "range": Array [
          90,
          90,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
    Object {
      "column": 22,
      "endColumn": 23,
      "endLine": 5,
      "fix": Object {
        "range": Array [
          54,
          54,
        ],
        "text": "($thing: String)",
      },
      "line": 5,
      "message": "Variable \\"$thing\\" is not defined by operation \\"Thing\\".",
      "nodeType": null,
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing {
~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    optional(thing: $thing)
~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
                     ~    [Variable \\"$thing\\" is not defined by operation \\"Thing\\".]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`missing-variable-with-another-variable.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing($other: String, $thing: String) {
    optional(thing: $thing)
    other: optional(thing: $other)
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 8,
      "fix": Object {
        "range": Array [
          141,
          141,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
    Object {
      "column": 22,
      "endColumn": 23,
      "endLine": 5,
      "fix": Object {
        "range": Array [
          69,
          69,
        ],
        "text": ", $thing: String",
      },
      "line": 5,
      "message": "Variable \\"$thing\\" is not defined by operation \\"Thing\\".",
      "nodeType": null,
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing($other: String) {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    optional(thing: $thing)
~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
                     ~    [Variable \\"$thing\\" is not defined by operation \\"Thing\\".]
    other: optional(thing: $other)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`missing-variable-with-another-variable-with-comma-after.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

// prettier-ignore
gql\`
  query Thing($other: String, $thing: String,,) {
    optional(thing: $thing)
    other: optional(thing: $other)
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 9,
      "fix": Object {
        "range": Array [
          162,
          162,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 4,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
    Object {
      "column": 22,
      "endColumn": 23,
      "endLine": 6,
      "fix": Object {
        "range": Array [
          88,
          88,
        ],
        "text": ", $thing: String",
      },
      "line": 6,
      "message": "Variable \\"$thing\\" is not defined by operation \\"Thing\\".",
      "nodeType": null,
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

// prettier-ignore
gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing($other: String,,) {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    optional(thing: $thing)
~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
                     ~    [Variable \\"$thing\\" is not defined by operation \\"Thing\\".]
    other: optional(thing: $other)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`multiple-validation-errors.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 5,
      "line": 5,
      "message": "Cannot query field \\"a\\" on type \\"Query\\".",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
    Object {
      "column": 5,
      "line": 6,
      "message": "Cannot query field \\"b\\" on type \\"Query\\".",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing {
    a
    ~    [Cannot query field \\"a\\" on type \\"Query\\".]
    b
~~~~~    [Cannot query field \\"a\\" on type \\"Query\\".]
    ~    [Cannot query field \\"b\\" on type \\"Query\\".]
  }
~~~    [Cannot query field \\"a\\" on type \\"Query\\".]
~~~    [Cannot query field \\"b\\" on type \\"Query\\".]
\`;
~~    [Cannot query field \\"a\\" on type \\"Query\\".]
~~    [Cannot query field \\"b\\" on type \\"Query\\".]

~    [Cannot query field \\"a\\" on type \\"Query\\".]
~    [Cannot query field \\"b\\" on type \\"Query\\".]",
}
`;

exports[`no-name.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "line": 3,
      "message": "GraphQL operations must have a name",
      "messageId": "mustBeNamed",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [GraphQL operations must have a name]
  query {
~~~~~~~~~    [GraphQL operations must have a name]
    something
~~~~~~~~~~~~~    [GraphQL operations must have a name]
  }
~~~    [GraphQL operations must have a name]
\`;
~    [GraphQL operations must have a name]
",
}
`;

exports[`parse-error.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 3,
      "line": 4,
      "message": "Syntax Error: Unexpected Name \\"que\\"",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
  que Thing {
  ~~~~~~~~~~~    [Syntax Error: Unexpected Name \\"que\\"]
    something
~~~~~~~~~~~~~    [Syntax Error: Unexpected Name \\"que\\"]
  }
~~~    [Syntax Error: Unexpected Name \\"que\\"]
\`;
~~    [Syntax Error: Unexpected Name \\"que\\"]

~    [Syntax Error: Unexpected Name \\"que\\"]",
}
`;

exports[`unknown-variable-type.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 23,
      "line": 4,
      "message": "Unknown type \\"DoesNotExist\\".",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing($thing: DoesNotExist) {
                      ~~~~~~~~~~~~~~~    [Unknown type \\"DoesNotExist\\".]
    optional(thing: $thing)
~~~~~~~~~~~~~~~~~~~~~~~~~~~    [Unknown type \\"DoesNotExist\\".]
  }
~~~    [Unknown type \\"DoesNotExist\\".]
\`;
~~    [Unknown type \\"DoesNotExist\\".]

~    [Unknown type \\"DoesNotExist\\".]",
}
`;

exports[`unused-variable.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing($thing: String) {
    optional
  }
\`as import(\\"../__generated__/ts-gql/Thing\\").type;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "fix": Object {
        "range": Array [
          91,
          91,
        ],
        "text": "as import(\\"../__generated__/ts-gql/Thing\\").type",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing($thing: String) {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    optional
~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`wrong-variable-type.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 15,
      "line": 4,
      "message": "Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "ts-gql",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
  query Thing($thing: ID) {
              ~~~~~~~~~~~~~    [Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".]
    optional(thing: $thing)
~~~~~~~~~~~~~~~~~~~~~~~~~~~    [Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".]
  }
~~~    [Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".]
\`;
~~    [Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".]

~    [Variable \\"$thing\\" of type \\"ID\\" used in position expecting type \\"String\\".]",
}
`;
